### DISCUSSION TITLE
Modbus TCP <-->RTU Gateway

### DISCUSSION THREAD

**Discussion by timaseth** [eModbus/eModbus#160] [Q&A] [answered]
Dear Devs,

I am new to the Arduino world and therefore, this is a bit of a struggle.

Aim: To create a ESP8266/ESP32 Wifi Gateway which will connect via GPIOs (using an appropriate RS485 to UART converter) to a Modbus RTU slave. Data polling will happen via another app over TCP port 502. For testing, I am using my old and trusted ModScan32.

Step 1: I downladed and extracted the 'src' folder to the Libraries folder of my Ard IDE.
Step 2: Installed ESP 32 board manager json using preferences.
Step 3: Copy pasted the https://github.com/eModbus/eModbus/tree/master/examples/WiFi-RTU_Bridge cpp in a new file of the IDE and attempted to compile.

Here is the error

Arduino: 1.8.19 (Windows Store 1.8.57.0) (Windows 10), Board: "ESP32 Wrover Module, Default 4MB with spiffs (1.2MB APP/1.5MB SPIFFS), QIO, 80MHz, 921600, None"

```
Multiple libraries were found for "WiFi.h"
 Used: C:\Users\Seth\Documents\ArduinoData\packages\esp32\hardware\esp32\1.0.6\libraries\WiFi
 Not used: C:\Program Files\WindowsApps\ArduinoLLC.ArduinoIDE_1.8.57.0_x86__mdqgnx93n4wtt\libraries\WiFi
In file included from C:\Users\Seth\Documents\Arduino\libraries\Modbus\ModbusClientTCPasync.cpp:5:0:
C:\Users\Seth\Documents\Arduino\libraries\Modbus\ModbusClientTCPasync.h:9:22: fatal error: AsyncTCP.h: No such file or directory
compilation terminated.
exit status 1
Error compiling for board ESP32 Wrover Module.
```
Questions
1. Is my Aim...well, practical and doable?
2. Where am I going wrong with the process steps.

Thanks in advance for your time and help!

**Comment by Miq1** [comment]
>    Is my Aim...well, practical and doable?

Of course it is. This is exactly what the ModbusBridge has been intended to do. You already picked the right example for that. 

>    Where am I going wrong with the process steps.

We are working in the PlatformIO environment, so at least I myself am lacking profound knowledge about the Arduino IDE. 
But the error message indicates that you are using the AsyncTCP variant, while WiFi.h belongs to the synchronous TCP version. You may look for some ``#include`` statements pulling in the one or the other and try to sort that out. 

Another guess: you might have to have the TCPasync library installed as well. The library manager may have it in stock.

**Reply by timaseth** [reply]
> 

Many thanks for such a quick reply - I use Platform IO to compile Tasmota (Home Automation firmware for ESP 8266). Do not see why I may not be able to read something onthe interwebs and try using PIO to compile emodbus.

Reverting soon...
PS: Will my 'aim' work on an esp8266?

**Reply by Miq1** [reply]
> PS: Will my 'aim' work on an esp8266?

Yes, most probably. I am using a ModbusServerTCPasync on several of my remote power plugs, that have an ESP8285. You will want to use the AsyncTCP version there - comrade Bert always is in favour of that on the ESP8266.

**Reply by bertmelis** [reply]
May I add, that on ESP32 using async or sync does not make that much of a difference. ESP32 runs on an OS and AsyncTCP is actually a sort of synchronous process using a separate task. On ESP8266 async is using direct callbacks from the LWIP stack.

So on ESP8266 you will see an improvement in performance, on ESP32 not so.

**Comment by timaseth** [comment]
Some progress!

Using Arduino, I was able to compile and upload to an esp32.
![image](https://user-images.githubusercontent.com/60132339/154263014-b0f9f88c-b668-4e65-91d3-493ae75ddc75.png)

However, I get an immediate disconnection from my modscan application when I try to poll.
Serial2 settings appear okay  - Serial2.begin(9600, SERIAL_8N1, GPIO_NUM_17, GPIO_NUM_16).
So does the Modbus address of the RTU slave (1 in my case)  MBbridge.attachServer(4, 1, ANY_FUNCTION_CODE, &MB);
![image](https://user-images.githubusercontent.com/60132339/154263315-cef1610b-76b9-4a5d-a822-4a7884ec497b.png)
![image](https://user-images.githubusercontent.com/60132339/154263871-5ef2cb87-40fe-4850-814d-229cbd195a54.png)

In my experince, something is very raidly closing the connection and iving modscan error code.

**Reply by Miq1** [reply]
The connection is closed if there is no more request within the configurable timeout period. IIRC this is 20 seconds by default, but may be changed with the ``start()`` call. This is done in the example sketch, where the timeout is set to a mere 600ms.

The server will accept the number of configured clients at a time only, so the timeout is meant to save resources. If you got only one client requesting, you may set the timeout to a huge value to prevent the disconnects completely.

**Reply by timaseth** [reply]
Great - I put in 10000 instead of the 600 and the connection remains stable. However, ever after polling the server, my device RX/TX lights never light up and therefore, no data.
![image](https://user-images.githubusercontent.com/60132339/154418501-44261b8b-81df-4ef2-87e2-37834d94739f.png)
My converter and dupont cables are known good - I have double checked them. 
RX/TX pins as per example skectch.
The only time I see tx/RX lights blink (that too, once) is when I reboot the device.

**Reply by Miq1** [reply]
Good that you are making a progress.

The first thing I thought when seeing your setup was: uh, no separate power for the adapter? But that may work well, I simply always have a dedicated PSU supplying all my components. 

Do you have a scope or logic analyzer at hand? Would be interesting to see if there is data on the TX2 pin at all. You may not see the LED light up, btw, as the data packets are so short. 

The TX2 pin (yellow cable) seems to be connected to one of the inner two pins on the adaptor. The adaptors I have do have their DI pin on the far right of the four pins, the inner being the DE and RE toggle pins. Is that different on yours?
[correction] You seem to have an auto halfduplex module, so the cabling looks okay.

**Reply by timaseth** [reply]
Thanks.
So, Tasmota firmware has a modbus poll option (not as a gateway but using a fixed number of Holding registers) which I run using the same RS485 to Uart converter and an ESP8266.
The RX TX data lights blink and I might add, blink very well.
All I am swapping from the Tasmota setup is using the eModbus flashed flashed esp32 - perhaps this is the reason I am confident that my modbus module and the RTU slave are all set up correctly (9600, 8N1, RTU slave address 1)
I am afraid I do not have a scope / data analyser.

This is what I have going (extension .txt added to upload)
[modServer.ino.txt](https://github.com/eModbus/eModbus/files/8086889/modServer.ino.txt)

Thanks again for all your help.

**Reply by Miq1** [reply]
Erm, you are using the GPIOs 17/16 for Serial2 in your code, but the wires are connected to RX2/TX2 on your board, which supposedly have GPIO numbers 25/27. Please note that I am using a DevkitCV4 board, which is different.
You may leave out the two GPIO arguments completely from the ``Serial2.begin()`` call to have the standard pins being used AFAIK.

**Reply by timaseth** [reply]
So, I tried removing the pin assignments and then tested all 3 UART pairs of the 32 - nothing.

I am back to my original code (worried that I might have made more changes than I should).

I was able to borrow a logic analyzer from my friend and that shows flat lines on the designated GPIOs (which i have checked against my board model and are good).

![image](https://user-images.githubusercontent.com/60132339/154483329-6fd40712-3106-4388-8ebb-211f39f44d05.png)

When I connect my tasmota modbus setup, this is the result
![image](https://user-images.githubusercontent.com/60132339/154483740-0b3503be-6509-4c19-a545-99052f8684b8.png)

As I suspected, there is no data TX RX happening.

**Reply by Miq1** [reply]
Strange indeed. We will need to tackle that from a different angle. Are you able to compile your sketch with the ``-DLOG_LEVEL=6`` flag and post the log here? The verbose log may show us what is happening.

**Reply by bertmelis** [reply]
For the record, are you compiling using Platformio or using Arduino IDE? In other words, what is the Arduino Core version you're using?

**Reply by timaseth** [reply]
I am using Arduino IDE 
![image](https://user-images.githubusercontent.com/60132339/154661115-601ecf94-065f-4f0d-b7db-1b0f2f80017f.png)

`-DLOG_LEVEL=6`
Where would I be putting this line? Sorry, like I said, I am new to the Arduino world.

**Reply by Miq1** [reply]
So you are using the Arduino IDE... It is a bit more complicated there to add compiler flags, so I would suggest you temporarily may change the library's ``Logging.h`` file for it. There is a line like ``#define LOG_LEVEL LOG_LEVEL_ERROR`` in it where you could change the latter to ``LOG_LEVEL_VERBOSE`` instead.
Once you do not need the complete logging any more, please be sure to undo the change.

**Comment by timaseth** [comment]
Update:

I was able to compile in PIO as well but face the exact same issue!
![image](https://user-images.githubusercontent.com/60132339/154668988-14283d2f-9c98-46dc-a900-9494c432e9b5.png)
![image](https://user-images.githubusercontent.com/60132339/154669051-1d6a49c2-7eda-4e56-aebe-bb85b0146756.png)

**Reply by Miq1** [reply]
No wonder - compilation in another IDE wil not change your code, obviously.

**Comment by timaseth** [comment]
Please tell me where to add which line in PIO and I shall share my log results - but explain as you would to a 12 year old child  ;-)

**Reply by Miq1** [reply]
😄
You will have a file called ``platformio.ini``. If you open it, there already may be a ``build_flags =`` line - if not, just add it to the ``[env] `` section of it, like ``build_flags = -DLOG_LEVEL=6``. If there is more than one build flag, you also may write one line as ``build_flags =`` and add the flags one per line below that, but indented by 2 space characters each.

**Comment by timaseth** [comment]
Here we go!
Once again, thank you for looking into this!
[modbuslog.txt](https://github.com/eModbus/eModbus/files/8096845/modbuslog.txt)

**Reply by Miq1** [reply]
For whatever reason the log still is not complete, but at least we see the ``E1`` error responses. E1 means "invalid server": you are requesting server ID 1, whereas the bridge is configured to listen as server ID 4 and 5 only.
There are the ``attachServer`` lines in the sketch that are defining which server ID on the listening side (TCP in your case) will be translated into what server ID on the requesting side (RTU in your code).
you must not ignore the error responses.. .

**Comment by timaseth** [comment]
Here is the .cpp file server part
```
// Define and start WiFi bridge

// ServerID 4: Server with remote serverID 1, accessed through RTU client MB
//             All FCs accepted, with the exception of FC 06
  MBbridge.attachServer(1, 1, ANY_FUNCTION_CODE, &MB);

// ServerID 5: Server with remote serverID 4, accessed through RTU client MB
//             Only FCs 03 and 04 accepted
  MBbridge.attachServer(1, 1, READ_HOLD_REGISTER, &MB);
  MBbridge.addFunctionCode(1, READ_INPUT_REGISTER);
```

[Log1.txt](https://github.com/eModbus/eModbus/files/8097417/Log1.txt)

**Reply by Miq1** [reply]
Two things: 

- the first ``attachServer`` call already opens the bridge for listening to ``ANY_FUNCTION_CODE`` and forwarding whatever request to the attached server. The following two ``attachServer`` and ``addFunctionCode`` calls are meaningless in that context. (The comments from the original example sketch may get misleading after a while...)
- The log shows an ``E0`` error now being reported - TIMEOUT. your RTU server with the ID 01 seems unresponsible.

**Comment by timaseth** [comment, marked as answer]
So, stubborn as I am (_or rather, as told by my wife!_), I decided to re-do everything.
Step 1. Modbus RTU Slave (address 01) device connected to a TCP enabled HMI device and confirmed working.
Step 2. Replaced the Modbus UART adapter and all assosiated wires.
Step 3. Took a fresh ESP32 rover module and flashed the code I was using last.

BOOM - Everything works from the word GO!!
Almost 7000+ polls made and not a single miss. Plus my adapter RX TX ligts burn as I am used to seeing them burn.

All I can say is **Thank You** for your guidance and your patience!
And for your Project!
![thank_you_PNG66](https://user-images.githubusercontent.com/60132339/155080773-ccf442ca-c76c-47d7-b839-408e1b2d169b.png)

**Reply by timaseth** [reply]
Quick Q:
Can the gateway run on an esp8266? If yes, what changes must I make. Thanks in advance.

**Reply by Miq1** [reply]
Not that easy, I am afraid. The ESP8266 basically has only one task, with the exception of the WiFi core. So it *could* be made working if you took the TCP async server code and would develop a non-tasking RTU client to continuously run in ``loop() `` - for the cost of timing difficulties. I do not consider this worthwhile, though, given the low prices for the ESP32.
